### ISSUE TITLE
Switch all methods to use a single options object

### ISSUE THREAD

**Issue by patrickarlt** [Esri/arcgis-rest-js#75]
Between https://github.com/Esri/arcgis-rest-js/issues/73 and https://github.com/Esri/arcgis-rest-js/pull/74 it seems we are struggling to figure out the right order of options and how to pass commonly used things like `owner` and `portal`. I am going to suggest the following:

1. Every method accepts a single `options` object.
2. That object is strictly typed and inherits from `IRequestOptions` or another options interface that inherits from `IRequestOptions`.
3. `request` now looks like
    ```ts
    interface IRequestOptions () {
      url: string;
      params?: IParams;
      httpMethod?: HTTPMethods;
      authentication?: IAuthenticationManager;
      fetch?: (input: RequestInfo, init?: RequestInit) => Promise<Response>;
    }

    request(options: IRequestOptions) {
      /** request **/
    }
    ```
4. Methods that wrap `request` are now implemented like:
    ```ts
    interface IReassignItemOptions extends IRequestOptions() {
      /**
       * `url` is not allowed in `IReassignItemOptions`. Use `portal` instead.
       */
      url: void;
      itemId: string;
      owner: string;
      newOwner: string;
      newFolder?: string;
      portal?: string;
    }

    reassignItem(options: IReassignItemOptions) {
      /*
       * `portal` and `currentOwner` can default to `options.authentication.username`
       * and `options.authentication.portal` or be overridden by the `portal` and `owner` options.
       */
    }
    ```
5. `portal` should be removed from `IRequestOptions` since not all methods require it https://github.com/Esri/arcgis-rest-js/blob/master/packages/arcgis-rest-request/src/request.ts#L51.

**Comment by dmfenton** [comment]
Strongly in favor

**Comment by tomwayson** [comment]
Why not keep `url` as the required first parameter of `request()` and out of `IRequestOptions` Seems like most wrapping fn's will end up having to `void` it.

**Comment by dbouwman** [comment]
Verbose in that almost every method needs it's own `IMethodOptions`, but at least it's consistent and we're not hiding *some* params in the request options hash.

**Comment by jgravois** [comment]
i'm good with anything. when an agreement has been reached, i will...

![make it so](https://vignette.wikia.nocookie.net/cardfight/images/9/97/Star-trek-make-it-so-450x270.jpg/revision/latest?cb=20160325065605)

**Comment by patrickarlt** [comment]
I'm all for what @tomwayson said about keeping URL as its own param in https://github.com/Esri/arcgis-rest-js/issues/75#issuecomment-349775514. I updated my original proposal to match.

**Comment by jgravois** [comment]
resolved via #78
